libs/ui/input/helm/src/lib/hlm-input.directive.ts

```diff
@@ -49,4 +58,36 @@ export class HlmInputDirective {
 	protected _computedClass = computed(() =>
 		hlm(inputVariants({ size: this._size(), error: this._error() }), this.userClass()),
 	);
+
+	ngControl: NgControl | null = this.injector.get(NgControl, null);
+
+	errorStateTracker: ErrorStateTracker;
+
+	private defaultErrorStateMatcher = inject(ErrorStateMatcher);
+	private parentForm = inject(NgForm, { optional: true });
+	private parentFormGroup = inject(FormGroupDirective, { optional: true });
+
+	errorState = computed(() => this.errorStateTracker.errorState());
+
+	constructor(@Inject(Injector) private injector: Injector) {
+		this.errorStateTracker = new ErrorStateTracker(
+			this.defaultErrorStateMatcher,
+			this.ngControl,
+			this.parentFormGroup,
+			this.parentForm,
+		);
+
+		effect(
+			() => {
+				if (this.ngControl) {
+					this.error = this.errorStateTracker.errorState();
+				}
+			},
+			{ allowSignalWrites: true },
+		);
+	}
+
+	ngDoCheck() {
```

Comment:
Why is this necessary? It feels like an ant-ipattern to have to do this in every supported component.

Reply:
I was seeing that material actually does it this way as well. They have an ngDoCheck in their form components and it's usually to call some sort of function to update the error state. One example from their repo https://github.com/angular/components/blob/ca7c090fb948c2f27f13d76e7cb567d0ba870ab9/src/material/select/select.ts#L678 . Otherwise not sure how else we could it.